consensus: do not assume the anchor point is genesis in prevPointAndBlockNo

nfrisby opened the pull request:
Fixes the first item in #1511.

Unfortunately I don't have a test case here, because the only repro I have only fails on my WIP branch for Issue 1489. On that branch, this commit does fix the repro.

I'm preparing an omnibus PR for 1489, but I thought I'd open this one too, since the fix seems clear in hindsight (unless I'm overlooking some invariant about the current chain here?) and it's in the library not just the test infrastructure.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -359,8 +359,10 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
+          (cc, cBno) <- lift $ atomically $
+            (,) <$> ChainDB.getCurrentChain chainDB
+                <*> ChainDB.getTipBlockNo chainDB
+          let mPrev = prevPointAndBlockNo currentSlot cBno cc
```
What about the following:
```haskell
          mPrev <- lift $ atomically $
            prevPointAndBlockNo currentSlot
                <$> ChainDB.getTipBlockNo chainDB
                <*> ChainDB.getCurrentChain chainDB
```
Because of laziness, we won't be doing the extra work in the transaction.

edsko replied:
Yes, let's do this.

edsko commented:
> There must have been a maximal rollback (so that we have an empty fragment not anchored at genesis) before we try to forge a block. Nice find! +1

Hold on a second. Does this mean switch-to-fork as rollback-followed-by-foll-forward is observable, and we might produce a block after the rollback but before the roll-forward? That can't happen, can it? We just add those blocks to the chain DB, and it picks a different chain. Why would block forging be able to observe a rollback (as opposed to a switch-to-fork)?

mrBliss commented:
> > There must have been a maximal rollback (so that we have an empty fragment not anchored at genesis) before we try to forge a block. Nice find! +1
> 
> Hold on a second. Does this mean switch-to-fork as rollback-followed-by-foll-forward is observable, and we might produce a block after the rollback but before the roll-forward? That can't happen, can it? We just add those blocks to the chain DB, and it picks a different chain. Why would block forging be able to observe a rollback (as opposed to a switch-to-fork)?

Hmm, yeah, that would be very strange :thinking:. @nfrisby Can you give some more details about the scenario in which this happens?

nfrisby commented:
> Hmm, yeah, that would be very strange 🤔. @nfrisby Can you give some more details about the scenario in which this happens?

I'll dig in a bit and report back. As of yet, all I retained was `k = 1`.

nfrisby commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -538,11 +541,22 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
+    mkPredecessor prevP bno
+        -- There is no block before it, so use genesis.
+      | 0 == bno  = (genesisPoint, genesisBlockNo)
+      | otherwise = (castPoint prevP, pred bno)
```
Gah EBBs!

If @b@ (as defined by the Haddock on `mkPredecessor`) is an EBB, then `pred bno` is wrong; it should be `bno` in that case.

nfrisby commented:
> So I think the only adjustment would be to change the outer `Empty` alternative to assert that the anchor is genesis. Agreed? (I suppose that's the one Edsko had in mind when he raised his objection.)

Hmm. Could there be a scenario where the ChainDB was truncated by exactly `k` blocks? Or would that cause a restart/reopen/etc such that the `getCurrentChain` `AnchorFragment` would still be non-`Empty`?

mrBliss commented:
> > So I think the only adjustment would be to change the outer `Empty` alternative to assert that the anchor is genesis. Agreed? (I suppose that's the one Edsko had in mind when he raised his objection.)
> 
> Hmm. Could there be a scenario where the ChainDB was truncated by exactly `k` blocks? Or would that cause a restart/reopen/etc such that the `getCurrentChain` `AnchorFragment` would still be non-`Empty`?

The chain of the ChainDB will never shrink while running, only when starting up. For example, when the whole VolatileDB folder is missing, the ChainDB can start with an empty fragment that is anchored at a point that is not genesis.

mrBliss commented:
I think this can be merged as-is or with [my suggestion](https://github.com/input-output-hk/ouroboros-network/pull/1544/files#r373366607).

edsko commented:
Ok, so, this can happen without EBBs or node restarts. Let `k = 1`. This means that under normal circumstances our fragment will be 1 block long. Now if another node manages to produce a block and send it to us before we manage to produce our own, we strip off the one block from our fragment and are left with an empty fragment. We should make sure that this scenario can occur in tests (#1569).

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -498,10 +500,11 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
-    prevPointAndBlockNo slot c = case c of
-        Empty _   -> Right (genesisPoint, genesisBlockNo)
+    prevPointAndBlockNo slot bno c = case c of
+        Empty p   -> Right $ mkPredecessor p bno
```
```suggestion
        Empty p   -> Right (castPoint p, succ bno)
```

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -538,11 +541,22 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
-             -> Right (genesisPoint, genesisBlockNo)
+             -> Right $ case c' of
+               _ :> hdr' -> (headerPoint hdr', blockNo hdr')
+               Empty p   -> mkPredecessor p (blockNo hdr)
```
```suggestion
               Empty p   -> (castPoint p, blockNo hdr)
```
And we won't even need `mkPredecessor`

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -498,10 +500,11 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
     prevPointAndBlockNo :: SlotNo
+                        -> BlockNo
                         -> AnchoredFragment (Header blk)
                         -> Either SlotNo (Point blk, BlockNo)
```
Let's change this `BlockNo` into the block number of the block *we're going to produce*, not the block number of the predecessor, because *there might not be a predecessor*. The docstring needs to be updated and the name of this function too.

mrBliss replied:
Instead of returning the slot from the future as `Left`, return the actual error, since we can fail not only because of a slot from the future, but also when there's already a block in the slot we're trying to produce a block for, and, moreover, that block is already immutable. This might happen in two cases: 1) the clock moved backwards, on restart we ignored everything from the VolatileDB since it's all in the future, and now the tip of the ImmutableDB (= anchor point of our empty chain fragment) points to a block produced in the same slot we're trying to produce a block in; 2) k = 0 and we already adopted a block from another leader of the same slot.

So introduce an actual error type for these two cases (use the above as part of its docstring) and add a new trace message for the extra case.

edsko commented:
So discussed this in detail with @mrBliss . We left concrete suggestions, but here's what could go wrong with the PR as it stands:

1. If the fragment is empty (say, due to data corruption and an empty volatile DB), then `bno` will be the tip of the immutable DB. That `bno` should then _be_ the block number of the predecessor; we should try to take _its_ predecessor (which is what currently happens).

2. If the fragment is empty after stripping off a block, and we reach genesis, `mkPredecessor` must produce a block number for the predecessor, _but there is none_, and so it has little choice but to return `genesisBlockNo`. However, that is wrong: when we then call `succ` later, we will have _skipped_ a block number.

I think the suggestions we made above will address both of these issues.

mrBliss commented:
I have created #1570 for tracking the bugs.

nfrisby commented:
I rewrote it and rebased onto origin/master. The diff is big, but I think it's clearer this way.

I am seeing one failure in RealPBFT when I cherry-pick this onto my omnibus PR for Issue 1489. I'm not sure if it's due to this, but it might be. On the other hand, BFT, LeaderSchedule, PBFT, and DualPBFT all passed 40000 tests. The error involves `ChainValidationDelegationSchedulingError`, so there's plenty of other moving parts besides this PR that could be responsible for it.

I unfortunately need to prepare for the meeting in the morning, so I'm stopping on this for now.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/Node/Tracers.hs:
```diff
@@ -177,6 +179,24 @@ data TraceForgeEvent blk tx
   -- See also <https://github.com/input-output-hk/ouroboros-network/issues/1462>
   | TraceBlockFromFuture SlotNo SlotNo
 
+  -- | Leadership check failed: the tip of the ImmDB inhabits a slot /after/
```
Not "after the current slot" but "the current slot". The ImmutableDB truncates blocks from the future, but not from the present (= the current slot), see #1563.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -553,6 +504,95 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
 newtype PRNG = PRNG ChaChaDRG
   deriving NoUnexpectedThunks via UseIsNormalForm PRNG
 
+-- | Data about a block's position in the chain
```
Mention that it's *about a block to be forged*, not an existing block. Otherwise the `succ` on L525 is a bit confusing.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -553,6 +504,95 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
+     -- If the fragment is 'Empty', then this is the block number of the
+     -- tip of the ImmDB.
+  -> Either (TraceForgeEvent blk (GenTx blk)) (BlockContext blk)
+     -- ^ the event is records the cause of the failure
```
```suggestion
     -- ^ the event records the cause of the failure
```

edsko commented:
I'll take care of these minor changes.

edsko commented:
bors r+
